pytorch_lightning/trainer/trainer.py

```diff
@@ -316,8 +316,61 @@ def __init__(
             # https://github.com/PyTorchLightning/pytorch-lightning/pull/1572/files#r420279383
             os.environ["HOROVOD_FUSION_THRESHOLD"] = str(0)
 
-        # Init callbacks
+        # training bookeeping
+        self.total_batch_idx = 0
+        self.running_loss = TensorRunningAccum(window_length=20)
+        self.batch_idx = 0
+        self.progress_bar_metrics = {}
+        self.callback_metrics = {}
+        self.num_val_batches = 0
+        self.num_training_batches = 0
+        self.num_test_batches = 0
+        self.train_dataloader = None
+        self.test_dataloaders = None
+        self.val_dataloaders = None
+
+        # training state
+        self.model = None
+        self.testing = False
+        self.disable_validation = False
+        self.lr_schedulers = []
+        self.optimizers = None
+        self.optimizer_frequencies = []
+        self.global_step = 0
+        self.current_epoch = 0
+        self.interrupted = False
+        self.should_stop = False
+
+        # set default save path if user didn't provide one
+        if default_root_dir is None or default_root_dir is True:
+            # Backward compatibility, TODO: remove default_save_path in v0.8.0
+            default_root_dir = default_save_path or os.getcwd()
+        self.default_root_dir = default_root_dir
+
+        # configure logger
+        self.configure_logger(logger)
+
+        # initialize callbacks
         self.callbacks = callbacks or []
+
+        # configure early stop callback
+        # creates a default one if none passed in
+        early_stop_callback = self.configure_early_stopping(early_stop_callback)
+        if early_stop_callback:
+            self.callbacks.append(early_stop_callback)
+
+        # configure checkpoint callback
+        # it is important that this is the last callback to run
+        # pass through the required args to figure out defaults
+        self.weights_save_path = weights_save_path
+        checkpoint_callback = self.configure_checkpoint_callback(checkpoint_callback)
+        if checkpoint_callback:
+            self.callbacks.append(checkpoint_callback)
```

Comment:
what is the order here? all callbacks, then early stop and the last checkpoint...
shall  be before early stop?

Reply:
checkpointing **_has_** to run after early stop so that the early stop callback state is stored properly. this is one of the things this PR aims to fix.